core/src/main/scala/kafka/log/Log.scala

```diff
@@ -565,19 +565,24 @@ class Log(val dir: File,
   def deleteOldSegments(predicate: LogSegment => Boolean): Int = {
     // find any segments that match the user-supplied predicate UNLESS it is the final segment
     // and it is empty (since we would just end up re-creating it
-    val lastSegment = activeSegment
-    val deletable = logSegments.takeWhile(s => predicate(s) && (s.baseOffset != lastSegment.baseOffset || s.size > 0))
-    val numToDelete = deletable.size
-    if(numToDelete > 0) {
```

Comment:
The aim here is to avoid synchronization for the case where `numToDelete == 0`. It seems like we should be able to avoid the NPE by changing `activeSegment` to return `null` if `lastEntry` is `null` and then acting accordingly here.